ICE: `interpret const eval failure of Unevaluated(UnevaluatedConst`

<!--
ICE: Rustc ./CDD080648C36199C7736BBD270BCC7818A8CAD62B140B0DE09163172D1551353.rs '' 'error: internal compiler error: compiler/rustc_const_eval/src/interpret/eval_context.rs:1195:33: interpret const eval failure of Unevaluated(UnevaluatedConst { def: DefId(0:9 ~ CDD080648C36199C7736BBD270BCC7818A8CAD62B140B0DE09163172D1551353[9ee3]::_), args: [], promoted: Some(promoted[0]) }, &'{erased} [Data; 0_usize]) which is not in required_consts', 'error: internal compiler error: compiler/rustc_const_eval/src/interpret/eval_context.rs:1195:33: interpret const eval failure of Unevaluated(UnevaluatedConst { def: DefId(0:9 ~ CDD080648C36199C7736BBD270BCC7818A8CAD62B140B0DE09163172D1551353[9ee3]::_), args: [], promoted: Some(promoted[0]) }, &'{erased} [Data; 0_usize]) which is not in required_consts'
File: /home/gh-matthiaskrgr/vcs/github/glacier2_2/o/CDD080648C36199C7736BBD270BCC7818A8CAD62B140B0DE09163172D1551353.rs
-->

latest manually reduced form:
```rust
pub struct Data([u8; usize::MAX >> 2]);
const _: &'static [Data] = &[];
```

auto-reduced (treereduce-rust):
````rust
pub struct Data([[[u32; 9999999]; 777777777]; 5_i32]);
const _: &'static Data = unsafe { &*(&[] as *const Data) };


````

original:
````rust
pub struct Data([[[u32; 9999999]; 777777777]; 5_i32]);
const _: &'static Data = unsafe { &*(&[] as *const Data) };

pub fn main() {}

````

Version information
````
rustc 1.80.0-nightly (78dd504f2 2024-05-24)
binary: rustc
commit-hash: 78dd504f2fd87c0cfabff7d9174253411caf2f80
commit-date: 2024-05-24
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.6

````

Command:
`/home/gh-matthiaskrgr/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.QyRvEWiVipgV/rustc_testrunner_tmpdir_reporting.6eRld1crLCV1/mvce.rs:2:60
  |
2 | const _: &'static Data = unsafe { &*(&[] as *const Data) };
  |                                                            ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.QyRvEWiVipgV/rustc_testrunner_tmpdir_reporting.6eRld1crLCV1/mvce.rs`

error[E0080]: evaluation of constant value failed
 --> /tmp/icemaker_global_tempdir.QyRvEWiVipgV/rustc_testrunner_tmpdir_reporting.6eRld1crLCV1/mvce.rs:2:39
  |
2 | const _: &'static Data = unsafe { &*(&[] as *const Data) };
  |                                       ^^ values of the type `[[u32; 9999999]; 777777777]` are too big for the current architecture

error: internal compiler error: compiler/rustc_const_eval/src/interpret/eval_context.rs:1195:33: interpret const eval failure of Unevaluated(UnevaluatedConst { def: DefId(0:9 ~ mvce[c32c]::_), args: [], promoted: Some(promoted[0]) }, &'{erased} [Data; 0_usize]) which is not in required_consts
 --> /tmp/icemaker_global_tempdir.QyRvEWiVipgV/rustc_testrunner_tmpdir_reporting.6eRld1crLCV1/mvce.rs:2:38
  |
2 | const _: &'static Data = unsafe { &*(&[] as *const Data) };
  |                                      ^^^

thread 'rustc' panicked at compiler/rustc_const_eval/src/interpret/eval_context.rs:1195:33:
Box<dyn Any>
stack backtrace:
   0:     0x7f43c6f8ca25 - std::backtrace_rs::backtrace::libunwind::trace::h8966c752ee3e4366
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x7f43c6f8ca25 - std::backtrace_rs::backtrace::trace_unsynchronized::h9ea5b98770d1aa33
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f43c6f8ca25 - std::sys_common::backtrace::_print_fmt::he48619da8989a3bf
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f43c6f8ca25 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h396d8ace60c94a52
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f43c6fdbafb - core::fmt::rt::Argument::fmt::h5a43b7685e1a8d8c
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/core/src/fmt/rt.rs:165:63
   5:     0x7f43c6fdbafb - core::fmt::write::hb0f90054e0ddacf7
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/core/src/fmt/mod.rs:1172:21
   6:     0x7f43c6f8177f - std::io::Write::write_fmt::hd596e75ff8fa0846
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/io/mod.rs:1835:15
   7:     0x7f43c6f8c7fe - std::sys_common::backtrace::_print::h40262024fa5f287e
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f43c6f8c7fe - std::sys_common::backtrace::print::ha7fcd41b8add839e
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f43c6f8f229 - std::panicking::default_hook::{{closure}}::h60e0c4e988d17123
  10:     0x7f43c6f8efca - std::panicking::default_hook::h9ed993c933fab0c2
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/panicking.rs:298:9
  11:     0x7f43ca114110 - std[4869c667286d786e]::panicking::update_hook::<alloc[803fe285b37d8cd0]::boxed::Box<rustc_driver_impl[957f6692bf6a3a58]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f43c6f8f95b - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hefe0f5c02c622009
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/alloc/src/boxed.rs:2077:9
  13:     0x7f43c6f8f95b - std::panicking::rust_panic_with_hook::h4ddf3c210c6e33af
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/panicking.rs:799:13
  14:     0x7f43ca1448d4 - std[4869c667286d786e]::panicking::begin_panic::<rustc_errors[9393b57c1e89acaf]::ExplicitBug>::{closure#0}
  15:     0x7f43ca1413b6 - std[4869c667286d786e]::sys_common::backtrace::__rust_end_short_backtrace::<std[4869c667286d786e]::panicking::begin_panic<rustc_errors[9393b57c1e89acaf]::ExplicitBug>::{closure#0}, !>
  16:     0x7f43ca13c756 - std[4869c667286d786e]::panicking::begin_panic::<rustc_errors[9393b57c1e89acaf]::ExplicitBug>
  17:     0x7f43ca14dd11 - <rustc_errors[9393b57c1e89acaf]::diagnostic::BugAbort as rustc_errors[9393b57c1e89acaf]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  18:     0x7f43ca071d78 - <rustc_errors[9393b57c1e89acaf]::DiagCtxt>::span_bug::<rustc_span[6eac4943cc7cd1a3]::span_encoding::Span, alloc[803fe285b37d8cd0]::string::String>
  19:     0x7f43ca0923fd - rustc_middle[34114187831b2967]::util::bug::opt_span_bug_fmt::<rustc_span[6eac4943cc7cd1a3]::span_encoding::Span>::{closure#0}
  20:     0x7f43ca09264a - rustc_middle[34114187831b2967]::ty::context::tls::with_opt::<rustc_middle[34114187831b2967]::util::bug::opt_span_bug_fmt<rustc_span[6eac4943cc7cd1a3]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f43ca087d6b - rustc_middle[34114187831b2967]::ty::context::tls::with_context_opt::<rustc_middle[34114187831b2967]::ty::context::tls::with_opt<rustc_middle[34114187831b2967]::util::bug::opt_span_bug_fmt<rustc_span[6eac4943cc7cd1a3]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f43c8c759a7 - rustc_middle[34114187831b2967]::util::bug::span_bug_fmt::<rustc_span[6eac4943cc7cd1a3]::span_encoding::Span>
  23:     0x7f43c9584bfe - <rustc_const_eval[2a16a84a09f3f441]::interpret::eval_context::InterpCx<rustc_const_eval[2a16a84a09f3f441]::const_eval::machine::CompileTimeInterpreter>>::eval_rvalue_into_place
  24:     0x7f43c950408c - rustc_const_eval[2a16a84a09f3f441]::const_eval::eval_queries::eval_to_allocation_raw_provider
  25:     0x7f43cbafa076 - rustc_query_impl[2cb4a958aeb3f22c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2cb4a958aeb3f22c]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[34114187831b2967]::query::erase::Erased<[u8; 24usize]>>
  26:     0x7f43cbaf9899 - rustc_query_system[5af415981e64bee5]::query::plumbing::try_execute_query::<rustc_query_impl[2cb4a958aeb3f22c]::DynamicConfig<rustc_query_system[5af415981e64bee5]::query::caches::DefaultCache<rustc_middle[34114187831b2967]::ty::ParamEnvAnd<rustc_middle[34114187831b2967]::mir::interpret::GlobalId>, rustc_middle[34114187831b2967]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[2cb4a958aeb3f22c]::plumbing::QueryCtxt, false>
  27:     0x7f43cbaf9470 - rustc_query_impl[2cb4a958aeb3f22c]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  28:     0x7f43cbaf8ff6 - rustc_middle[34114187831b2967]::query::plumbing::query_get_at::<rustc_query_system[5af415981e64bee5]::query::caches::DefaultCache<rustc_middle[34114187831b2967]::ty::ParamEnvAnd<rustc_middle[34114187831b2967]::mir::interpret::GlobalId>, rustc_middle[34114187831b2967]::query::erase::Erased<[u8; 24usize]>>>
  29:     0x7f43cbaf747c - rustc_const_eval[2a16a84a09f3f441]::const_eval::eval_queries::eval_to_const_value_raw_provider
  30:     0x7f43cbaf73f6 - rustc_query_impl[2cb4a958aeb3f22c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2cb4a958aeb3f22c]::query_impl::eval_to_const_value_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[34114187831b2967]::query::erase::Erased<[u8; 24usize]>>
  31:     0x7f43cbaf985c - rustc_query_system[5af415981e64bee5]::query::plumbing::try_execute_query::<rustc_query_impl[2cb4a958aeb3f22c]::DynamicConfig<rustc_query_system[5af415981e64bee5]::query::caches::DefaultCache<rustc_middle[34114187831b2967]::ty::ParamEnvAnd<rustc_middle[34114187831b2967]::mir::interpret::GlobalId>, rustc_middle[34114187831b2967]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[2cb4a958aeb3f22c]::plumbing::QueryCtxt, false>
  32:     0x7f43cbaf9370 - rustc_query_impl[2cb4a958aeb3f22c]::query_impl::eval_to_const_value_raw::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7f43cb754bdf - <rustc_middle[34114187831b2967]::hir::map::Map>::par_body_owners::<rustc_hir_analysis[513396f821121b45]::check_crate::{closure#3}>::{closure#0}
  34:     0x7f43cb753022 - rustc_hir_analysis[513396f821121b45]::check_crate
  35:     0x7f43cb73323e - rustc_interface[6c4077a70579b4d0]::passes::analysis
  36:     0x7f43cb732d95 - rustc_query_impl[2cb4a958aeb3f22c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2cb4a958aeb3f22c]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[34114187831b2967]::query::erase::Erased<[u8; 1usize]>>
  37:     0x7f43cc13a8a5 - rustc_query_system[5af415981e64bee5]::query::plumbing::try_execute_query::<rustc_query_impl[2cb4a958aeb3f22c]::DynamicConfig<rustc_query_system[5af415981e64bee5]::query::caches::SingleCache<rustc_middle[34114187831b2967]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[2cb4a958aeb3f22c]::plumbing::QueryCtxt, false>
  38:     0x7f43cc13a60f - rustc_query_impl[2cb4a958aeb3f22c]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  39:     0x7f43cbfb908e - rustc_interface[6c4077a70579b4d0]::interface::run_compiler::<core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>, rustc_driver_impl[957f6692bf6a3a58]::run_compiler::{closure#0}>::{closure#1}
  40:     0x7f43cbfa4927 - std[4869c667286d786e]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[6c4077a70579b4d0]::util::run_in_thread_with_globals<rustc_interface[6c4077a70579b4d0]::util::run_in_thread_pool_with_globals<rustc_interface[6c4077a70579b4d0]::interface::run_compiler<core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>, rustc_driver_impl[957f6692bf6a3a58]::run_compiler::{closure#0}>::{closure#1}, core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>>::{closure#0}, core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>>
  41:     0x7f43cbfa46ea - <<std[4869c667286d786e]::thread::Builder>::spawn_unchecked_<rustc_interface[6c4077a70579b4d0]::util::run_in_thread_with_globals<rustc_interface[6c4077a70579b4d0]::util::run_in_thread_pool_with_globals<rustc_interface[6c4077a70579b4d0]::interface::run_compiler<core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>, rustc_driver_impl[957f6692bf6a3a58]::run_compiler::{closure#0}>::{closure#1}, core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>>::{closure#0}, core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[cb4e43784437d8f1]::result::Result<(), rustc_span[6eac4943cc7cd1a3]::ErrorGuaranteed>>::{closure#2} as core[cb4e43784437d8f1]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  42:     0x7f43c6f9973b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h496e4fc5fb53ece2
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/alloc/src/boxed.rs:2063:9
  43:     0x7f43c6f9973b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hd4fa5d81a55c123f
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/alloc/src/boxed.rs:2063:9
  44:     0x7f43c6f9973b - std::sys::pal::unix::thread::Thread::new::thread_start::h04d31b4f40ca136c
                               at /rustc/78dd504f2fd87c0cfabff7d9174253411caf2f80/library/std/src/sys/pal/unix/thread.rs:108:17
  45:     0x7f43c6c94ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  46:     0x7f43c6d26850 - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  47:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.80.0-nightly (78dd504f2 2024-05-24) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [eval_to_allocation_raw] const-evaluating + checking `_`
#1 [eval_to_const_value_raw] simplifying constant for the type system `_`
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0080, E0601.
For more information about an error, try `rustc --explain E0080`.

```

</p>
</details>

<!--
query stack:
2 | const _: &'static Data = unsafe { &*(&[] as *const Data) };
2 | const _: &'static Data = unsafe { &*(&[] as *const Data) };
2 | const _: &'static Data = unsafe { &*(&[] as *const Data) };
#0 [eval_to_allocation_raw] const-evaluating + checking `_`
#1 [eval_to_const_value_raw] simplifying constant for the type system `_`

-->

--- a/compiler/rustc_const_eval/src/const_eval/error.rs
+++ b/compiler/rustc_const_eval/src/const_eval/error.rs
@@ -139,12 +139,14 @@ where
     match error {
         // Don't emit a new diagnostic for these errors, they are already reported elsewhere or
         // should remain silent.
+        err_inval!(AlreadyReported(info)) => ErrorHandled::Reported(info, span),
         err_inval!(Layout(LayoutError::Unknown(_))) | err_inval!(TooGeneric) => {
             ErrorHandled::TooGeneric(span)
         }
-        err_inval!(AlreadyReported(guar)) => ErrorHandled::Reported(guar, span),
         err_inval!(Layout(LayoutError::ReferencesError(guar))) => {
-            ErrorHandled::Reported(ReportedErrorInfo::tainted_by_errors(guar), span)
+            // This can occur in infallible promoteds e.g. when a non-existent type or field is
+            // encountered.
+            ErrorHandled::Reported(ReportedErrorInfo::allowed_in_infallible(guar), span)
         }
         // Report remaining errors.
         _ => {
@@ -152,7 +154,12 @@ where
             let span = span.substitute_dummy(our_span);
             let err = mk(span, frames);
             let mut err = tcx.dcx().create_err(err);
-            let can_be_spurious = matches!(error, InterpErrorKind::ResourceExhaustion(_));
+            // We allow invalid programs in infallible promoteds since invalid layouts can occur
+            // anyway (e.g. due to size overflow). And we allow OOM as that can happen any time.
+            let allowed_in_infallible = matches!(
+                error,
+                InterpErrorKind::ResourceExhaustion(_) | InterpErrorKind::InvalidProgram(_)
+            );
 
             let msg = error.diagnostic_message();
             error.add_args(&mut err);
@@ -160,8 +167,8 @@ where
             // Use *our* span to label the interp error
             err.span_label(our_span, msg);
             let g = err.emit();
-            let reported = if can_be_spurious {
-                ReportedErrorInfo::spurious(g)
+            let reported = if allowed_in_infallible {
+                ReportedErrorInfo::allowed_in_infallible(g)
             } else {
                 ReportedErrorInfo::from(g)
             };

--- a/compiler/rustc_const_eval/src/interpret/eval_context.rs
+++ b/compiler/rustc_const_eval/src/interpret/eval_context.rs
@@ -268,7 +268,7 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         };
         // do not continue if typeck errors occurred (can only occur in local crate)
         if let Some(err) = body.tainted_by_errors {
-            throw_inval!(AlreadyReported(ReportedErrorInfo::tainted_by_errors(err)));
+            throw_inval!(AlreadyReported(ReportedErrorInfo::from(err)));
         }
         interp_ok(body)
     }
@@ -585,13 +585,10 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                     match err {
                         ErrorHandled::TooGeneric(..) => {},
                         ErrorHandled::Reported(reported, span) => {
-                            if reported.is_tainted_by_errors() {
-                                // const-eval will return "tainted" errors if e.g. the layout cannot
-                                // be computed as the type references non-existing names.
-                                // See <https://github.com/rust-lang/rust/issues/124348>.
-                            } else if reported.can_be_spurious() {
+                            if reported.is_allowed_in_infallible() {
                                 // These errors can just sometimes happen, even when the expression
-                                // is nominally "infallible", e.g. when running out of memory.
+                                // is nominally "infallible", e.g. when running out of memory
+                                // or when some layout could not be computed.
                             } else {
                                 // Looks like the const is not captured by `required_consts`, that's bad.
                                 span_bug!(span, "interpret const eval failure of {val:?} which is not in required_consts");

--- a/compiler/rustc_middle/src/mir/interpret/error.rs
+++ b/compiler/rustc_middle/src/mir/interpret/error.rs
@@ -46,7 +46,7 @@ impl ErrorHandled {
     pub fn emit_note(&self, tcx: TyCtxt<'_>) {
         match self {
             &ErrorHandled::Reported(err, span) => {
-                if !err.is_tainted_by_errors && !span.is_dummy() {
+                if !err.allowed_in_infallible && !span.is_dummy() {
                     tcx.dcx().emit_note(error::ErroneousConstant { span });
                 }
             }
@@ -58,34 +58,26 @@ impl ErrorHandled {
 #[derive(Debug, Copy, Clone, PartialEq, Eq, HashStable, TyEncodable, TyDecodable)]
 pub struct ReportedErrorInfo {
     error: ErrorGuaranteed,
-    is_tainted_by_errors: bool,
-    /// Whether this is the kind of error that can sometimes occur, and sometimes not.
-    /// Used for resource exhaustion errors.
-    can_be_spurious: bool,
+    /// Whether this error is allowed to show up even in otherwise "infallible" promoteds.
+    /// This is for things like overflows during size computation or resource exhaustion.
+    allowed_in_infallible: bool,
 }
 
 impl ReportedErrorInfo {
     #[inline]
-    pub fn tainted_by_errors(error: ErrorGuaranteed) -> ReportedErrorInfo {
-        ReportedErrorInfo { is_tainted_by_errors: true, can_be_spurious: false, error }
-    }
-    #[inline]
-    pub fn spurious(error: ErrorGuaranteed) -> ReportedErrorInfo {
-        ReportedErrorInfo { can_be_spurious: true, is_tainted_by_errors: false, error }
+    pub fn allowed_in_infallible(error: ErrorGuaranteed) -> ReportedErrorInfo {
+        ReportedErrorInfo { allowed_in_infallible: true, error }
     }
 
-    pub fn is_tainted_by_errors(&self) -> bool {
-        self.is_tainted_by_errors
-    }
-    pub fn can_be_spurious(&self) -> bool {
-        self.can_be_spurious
+    pub fn is_allowed_in_infallible(&self) -> bool {
+        self.allowed_in_infallible
     }
 }
 
 impl From<ErrorGuaranteed> for ReportedErrorInfo {
     #[inline]
     fn from(error: ErrorGuaranteed) -> ReportedErrorInfo {
-        ReportedErrorInfo { is_tainted_by_errors: false, can_be_spurious: false, error }
+        ReportedErrorInfo { allowed_in_infallible: false, error }
     }
 }
 

--- a/tests/crashes/125476.rs
+++ b/tests/crashes/125476.rs
@@ -1,4 +1,0 @@
-//@ known-bug: rust-lang/rust#125476
-//@ only-x86_64
-pub struct Data([u8; usize::MAX >> 2]);
-const _: &'static [Data] = &[];

--- a/tests/ui/const-generics/generic_const_exprs/unevaluated-const-ice-119731.stderr
+++ b/tests/ui/const-generics/generic_const_exprs/unevaluated-const-ice-119731.stderr
@@ -72,6 +72,20 @@ help: add `#![feature(adt_const_params)]` to the crate attributes to enable more
 LL + #![feature(adt_const_params)]
    |
 
+note: erroneous constant encountered
+  --> $DIR/unevaluated-const-ice-119731.rs:22:19
+   |
+LL |     impl v17<512, v0> {
+   |                   ^^
+
+note: erroneous constant encountered
+  --> $DIR/unevaluated-const-ice-119731.rs:22:19
+   |
+LL |     impl v17<512, v0> {
+   |                   ^^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
 error: maximum number of nodes exceeded in constant v20::v17::<v10, v2>::{constant#0}
   --> $DIR/unevaluated-const-ice-119731.rs:28:37
    |

--- a/tests/ui/consts/const-integer-bool-ops.stderr
+++ b/tests/ui/consts/const-integer-bool-ops.stderr
@@ -16,6 +16,12 @@ error[E0308]: mismatched types
 LL | const X: usize = 42 && 39;
    |                  ^^^^^^^^ expected `usize`, found `bool`
 
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:8:18
+   |
+LL | const ARR: [i32; X] = [99; 34];
+   |                  ^
+
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:10:19
    |
@@ -33,6 +39,12 @@ error[E0308]: mismatched types
    |
 LL | const X1: usize = 42 || 39;
    |                   ^^^^^^^^ expected `usize`, found `bool`
+
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:17:19
+   |
+LL | const ARR1: [i32; X1] = [99; 47];
+   |                   ^^
 
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:19:19
@@ -52,6 +64,12 @@ error[E0308]: mismatched types
 LL | const X2: usize = -42 || -39;
    |                   ^^^^^^^^^^ expected `usize`, found `bool`
 
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:26:19
+   |
+LL | const ARR2: [i32; X2] = [99; 18446744073709551607];
+   |                   ^^
+
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:28:19
    |
@@ -70,11 +88,23 @@ error[E0308]: mismatched types
 LL | const X3: usize = -42 && -39;
    |                   ^^^^^^^^^^ expected `usize`, found `bool`
 
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:35:19
+   |
+LL | const ARR3: [i32; X3] = [99; 6];
+   |                   ^^
+
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:37:18
    |
 LL | const Y: usize = 42.0 == 42.0;
    |                  ^^^^^^^^^^^^ expected `usize`, found `bool`
+
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:40:19
+   |
+LL | const ARRR: [i32; Y] = [99; 1];
+   |                   ^
 
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:42:19
@@ -82,11 +112,23 @@ error[E0308]: mismatched types
 LL | const Y1: usize = 42.0 >= 42.0;
    |                   ^^^^^^^^^^^^ expected `usize`, found `bool`
 
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:45:20
+   |
+LL | const ARRR1: [i32; Y1] = [99; 1];
+   |                    ^^
+
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:47:19
    |
 LL | const Y2: usize = 42.0 <= 42.0;
    |                   ^^^^^^^^^^^^ expected `usize`, found `bool`
+
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:50:20
+   |
+LL | const ARRR2: [i32; Y2] = [99; 1];
+   |                    ^^
 
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:52:19
@@ -94,17 +136,35 @@ error[E0308]: mismatched types
 LL | const Y3: usize = 42.0 > 42.0;
    |                   ^^^^^^^^^^^ expected `usize`, found `bool`
 
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:55:20
+   |
+LL | const ARRR3: [i32; Y3] = [99; 0];
+   |                    ^^
+
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:57:19
    |
 LL | const Y4: usize = 42.0 < 42.0;
    |                   ^^^^^^^^^^^ expected `usize`, found `bool`
 
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:60:20
+   |
+LL | const ARRR4: [i32; Y4] = [99; 0];
+   |                    ^^
+
 error[E0308]: mismatched types
   --> $DIR/const-integer-bool-ops.rs:62:19
    |
 LL | const Y5: usize = 42.0 != 42.0;
    |                   ^^^^^^^^^^^^ expected `usize`, found `bool`
+
+note: erroneous constant encountered
+  --> $DIR/const-integer-bool-ops.rs:65:20
+   |
+LL | const ARRR5: [i32; Y5] = [99; 0];
+   |                    ^^
 
 error: aborting due to 18 previous errors
 

--- a/tests/ui/consts/const-mut-refs/issue-76510.stderr
+++ b/tests/ui/consts/const-mut-refs/issue-76510.stderr
@@ -4,6 +4,12 @@ error[E0764]: mutable references are not allowed in the final value of constants
 LL | const S: &'static mut str = &mut " hello ";
    |                             ^^^^^^^^^^^^^^
 
+note: erroneous constant encountered
+  --> $DIR/issue-76510.rs:7:70
+   |
+LL |         let s = transmute::<(*const u8, usize), &ManuallyDrop<str>>((S.as_ptr(), 3));
+   |                                                                      ^
+
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0764`.

--- a/tests/ui/consts/const-tup-index-span.stderr
+++ b/tests/ui/consts/const-tup-index-span.stderr
@@ -11,6 +11,12 @@ help: use a trailing comma to create a tuple with one element
 LL | const TUP: (usize,) = (5usize << 64,);
    |                       +            ++
 
+note: erroneous constant encountered
+  --> $DIR/const-tup-index-span.rs:6:18
+   |
+LL | const ARR: [i32; TUP.0] = [];
+   |                  ^^^
+
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/consts/issue-54954.stderr
+++ b/tests/ui/consts/issue-54954.stderr
@@ -19,6 +19,24 @@ LL | |         core::mem::size_of::<T>()
 LL | |     }
    | |_____- `Tt::const_val` defined here
 
+note: erroneous constant encountered
+  --> $DIR/issue-54954.rs:11:15
+   |
+LL | fn f(z: [f32; ARR_LEN]) -> [f32; ARR_LEN] {
+   |               ^^^^^^^
+
+note: erroneous constant encountered
+  --> $DIR/issue-54954.rs:11:34
+   |
+LL | fn f(z: [f32; ARR_LEN]) -> [f32; ARR_LEN] {
+   |                                  ^^^^^^^
+
+note: erroneous constant encountered
+  --> $DIR/issue-54954.rs:16:22
+   |
+LL |     let _ = f([1f32; ARR_LEN]);
+   |                      ^^^^^^^
+
 error: aborting due to 2 previous errors
 
 Some errors have detailed explanations: E0379, E0790.

--- a/tests/ui/consts/missing_assoc_const_type2.stderr
+++ b/tests/ui/consts/missing_assoc_const_type2.stderr
@@ -4,5 +4,11 @@ error: missing type for `const` item
 LL |     const FIRST:  = 10;
    |                 ^ help: provide a type for the associated constant: `u8`
 
+note: erroneous constant encountered
+  --> $DIR/missing_assoc_const_type2.rs:18:5
+   |
+LL |     TwoDigits::FIRST as usize
+   |     ^^^^^^^^^^^^^^^^
+
 error: aborting due to 1 previous error
 

--- a/tests/ui/consts/promoted_running_out_of_memory_issue-130687.stderr
+++ b/tests/ui/consts/promoted_running_out_of_memory_issue-130687.stderr
@@ -4,12 +4,6 @@ error[E0080]: evaluation of constant value failed
 LL | const _: &'static Data = &Data([0; (1 << 47) - 1]);
    |                                ^^^^^^^^^^^^^^^^^^ tried to allocate more memory than available to compiler
 
-note: erroneous constant encountered
-  --> $DIR/promoted_running_out_of_memory_issue-130687.rs:8:26
-   |
-LL | const _: &'static Data = &Data([0; (1 << 47) - 1]);
-   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^
-
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0080`.

--- a/tests/ui/consts/promoted_size_overflow.rs
+++ b/tests/ui/consts/promoted_size_overflow.rs
@@ -1,0 +1,7 @@
+//@ only-64bit
+pub struct Data([u8; usize::MAX >> 2]);
+const _: &'static [Data] = &[];
+//~^ERROR: evaluation of constant value failed
+//~| too big for the target architecture
+
+fn main() {}

--- a/tests/ui/consts/promoted_size_overflow.stderr
+++ b/tests/ui/consts/promoted_size_overflow.stderr
@@ -1,0 +1,9 @@
+error[E0080]: evaluation of constant value failed
+  --> $DIR/promoted_size_overflow.rs:3:29
+   |
+LL | const _: &'static [Data] = &[];
+   |                             ^^ values of the type `[u8; 4611686018427387903]` are too big for the target architecture
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0080`.

--- a/tests/ui/consts/uninhabited-const-issue-61744.rs
+++ b/tests/ui/consts/uninhabited-const-issue-61744.rs
@@ -5,15 +5,15 @@ pub const unsafe fn fake_type<T>() -> T {
 }
 
 pub const unsafe fn hint_unreachable() -> ! {
-    fake_type()
+    fake_type() //~ inside
 }
 
 trait Const {
-    const CONSTANT: i32 = unsafe { fake_type() };
+    const CONSTANT: i32 = unsafe { fake_type() }; //~ inside
 }
 
 impl<T> Const for T {}
 
 pub fn main() -> () {
-    dbg!(i32::CONSTANT); //~ constant
+    dbg!(i32::CONSTANT);
 }

--- a/tests/ui/consts/uninhabited-const-issue-61744.stderr
+++ b/tests/ui/consts/uninhabited-const-issue-61744.stderr
@@ -645,20 +645,6 @@ note: inside `<i32 as Const>::CONSTANT`
 LL |     const CONSTANT: i32 = unsafe { fake_type() };
    |                                    ^^^^^^^^^^^
 
-note: erroneous constant encountered
-  --> $DIR/uninhabited-const-issue-61744.rs:18:10
-   |
-LL |     dbg!(i32::CONSTANT);
-   |          ^^^^^^^^^^^^^
-
-note: erroneous constant encountered
-  --> $DIR/uninhabited-const-issue-61744.rs:18:10
-   |
-LL |     dbg!(i32::CONSTANT);
-   |          ^^^^^^^^^^^^^
-   |
-   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
-
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0080`.

--- a/tests/ui/enum-discriminant/issue-41394.stderr
+++ b/tests/ui/enum-discriminant/issue-41394.stderr
@@ -6,6 +6,12 @@ LL |     A = "" + 1
    |         |
    |         &str
 
+note: erroneous constant encountered
+  --> $DIR/issue-41394.rs:7:9
+   |
+LL |     A = Foo::A as isize
+   |         ^^^^^^^^^^^^^^^
+
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0369`.

--- a/tests/ui/explicit-tail-calls/ctfe-id-unlimited.return.stderr
+++ b/tests/ui/explicit-tail-calls/ctfe-id-unlimited.return.stderr
@@ -25,12 +25,6 @@ note: inside `ID_ED`
 LL | const ID_ED: u32 = rec_id(ORIGINAL);
    |                    ^^^^^^^^^^^^^^^^
 
-note: erroneous constant encountered
-  --> $DIR/ctfe-id-unlimited.rs:31:40
-   |
-LL | const ASSERT: () = assert!(ORIGINAL == ID_ED);
-   |                                        ^^^^^
-
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0080`.

--- a/tests/ui/layout/base-layout-is-sized-ice-123078.stderr
+++ b/tests/ui/layout/base-layout-is-sized-ice-123078.stderr
@@ -25,6 +25,12 @@ LL | const C: S = unsafe { std::mem::transmute(()) };
    = note: source type: `()` (0 bits)
    = note: target type: `S` (size can vary because of [u8])
 
+note: erroneous constant encountered
+  --> $DIR/base-layout-is-sized-ice-123078.rs:13:5
+   |
+LL |     C;
+   |     ^
+
 error: aborting due to 2 previous errors
 
 Some errors have detailed explanations: E0277, E0512.

--- a/tests/ui/limits/issue-55878.stderr
+++ b/tests/ui/limits/issue-55878.stderr
@@ -11,23 +11,6 @@ note: inside `main`
 LL |     println!("Size: {}", std::mem::size_of::<[u8; u64::MAX as usize]>());
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-note: erroneous constant encountered
-  --> $DIR/issue-55878.rs:7:26
-   |
-LL |     println!("Size: {}", std::mem::size_of::<[u8; u64::MAX as usize]>());
-   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-   |
-   = note: this note originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
-
-note: erroneous constant encountered
-  --> $DIR/issue-55878.rs:7:26
-   |
-LL |     println!("Size: {}", std::mem::size_of::<[u8; u64::MAX as usize]>());
-   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-   |
-   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
-   = note: this note originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
-
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0080`.

--- a/tests/ui/resolve/generic-params-from-outer-item-in-const-item.default.stderr
+++ b/tests/ui/resolve/generic-params-from-outer-item-in-const-item.default.stderr
@@ -29,6 +29,20 @@ LL |     const _: u32 = T::C;
    |
    = note: a `const` is a separate item from the item that contains it
 
+note: erroneous constant encountered
+  --> $DIR/generic-params-from-outer-item-in-const-item.rs:22:9
+   |
+LL |         I
+   |         ^
+
+note: erroneous constant encountered
+  --> $DIR/generic-params-from-outer-item-in-const-item.rs:22:9
+   |
+LL |         I
+   |         ^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
 error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0401`.

--- a/tests/ui/resolve/generic-params-from-outer-item-in-const-item.generic_const_items.stderr
+++ b/tests/ui/resolve/generic-params-from-outer-item-in-const-item.generic_const_items.stderr
@@ -35,6 +35,20 @@ LL |     const _: u32 = T::C;
    |
    = note: a `const` is a separate item from the item that contains it
 
+note: erroneous constant encountered
+  --> $DIR/generic-params-from-outer-item-in-const-item.rs:22:9
+   |
+LL |         I
+   |         ^
+
+note: erroneous constant encountered
+  --> $DIR/generic-params-from-outer-item-in-const-item.rs:22:9
+   |
+LL |         I
+   |         ^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
 error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0401`.

--- a/tests/ui/resolve/issue-50599.stderr
+++ b/tests/ui/resolve/issue-50599.stderr
@@ -20,6 +20,12 @@ LL -     const M: usize = (f64::from(N) * std::f64::LOG10_2) as usize;
 LL +     const M: usize = (f64::from(N) * LOG10_2) as usize;
    |
 
+note: erroneous constant encountered
+  --> $DIR/issue-50599.rs:4:29
+   |
+LL |     let mut digits = [0u32; M];
+   |                             ^
+
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0425`.

--- a/tests/ui/rfcs/rfc-2497-if-let-chains/avoid-invalid-mir.stderr
+++ b/tests/ui/rfcs/rfc-2497-if-let-chains/avoid-invalid-mir.stderr
@@ -6,5 +6,11 @@ LL |     !let y = 42;
    |
    = note: only supported directly in conditions of `if` and `while` expressions
 
+note: erroneous constant encountered
+  --> $DIR/avoid-invalid-mir.rs:11:13
+   |
+LL |     x: [(); N]
+   |             ^
+
 error: aborting due to 1 previous error
 

--- a/tests/ui/self/arbitrary-self-from-method-substs-ice.stderr
+++ b/tests/ui/self/arbitrary-self-from-method-substs-ice.stderr
@@ -17,6 +17,12 @@ LL |     const fn get<R: Deref<Target = Self>>(self: R) -> u32 {
 LL |     }
    |     - value is dropped here
 
+note: erroneous constant encountered
+  --> $DIR/arbitrary-self-from-method-substs-ice.rs:24:5
+   |
+LL |     FOO;
+   |     ^^^
+
 error[E0801]: invalid generic `self` parameter type: `R`
   --> $DIR/arbitrary-self-from-method-substs-ice.rs:10:49
    |

--- a/tests/ui/type/type-dependent-def-issue-49241.stderr
+++ b/tests/ui/type/type-dependent-def-issue-49241.stderr
@@ -9,6 +9,12 @@ help: consider using `let` instead of `const`
 LL |     let l: usize = v.count();
    |     ~~~
 
+note: erroneous constant encountered
+  --> $DIR/type-dependent-def-issue-49241.rs:4:18
+   |
+LL |     let s: [u32; l] = v.into_iter().collect();
+   |                  ^
+
 error: aborting due to 1 previous error
 
 For more information about this error, try `rustc --explain E0435`.

